src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/Helper.cs

```diff
@@ -0,0 +1,247 @@
+                {
+                    foreach (string subdirectory in Directory.GetDirectories(directory))
+                    {
+                        if (Path.GetDirectoryName(subdirectory)?.Length == 12)
+                        {
+                            return subdirectory;
+                        }
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private static Mutex CreateMutexNotOwned(string pathName)
+        {
+            return new Mutex(initiallyOwned: false, name: @"Global\" + GetStrongHashSuitableForObjectName(pathName));
+        }
+
+        internal static string GetNormalizedUriHash(Uri uri)
+        {
+            // On desktop System.Security.Url is used as evidence, it has an internal Normalize() method.
+            // Uri.ToString() appears to be functionally equivalent.
+            return GetStrongHashSuitableForObjectName(uri.ToString());
+        }
+
+        internal static string GetNormalizedStrongNameHash(AssemblyName name)
+        {
+            byte[] publicKey = name.GetPublicKey();
+
+            // If we don't have a key, we're not strong named
+            if (publicKey == null || publicKey.Length == 0)
+                return null;
+
+            // Emulate what we get from StrongName.Normalize().
+            using (MemoryStream ms = new MemoryStream())
+            {
+                using (BinaryWriter bw = new BinaryWriter(ms))
+                {
+                    bw.Write(publicKey);
+                    bw.Write(name.Version.Major);
+                    bw.Write(name.Name);
+                }
+
+                ms.Position = 0;
+                return GetStrongHashSuitableForObjectName(ms);
+            }
+        }
+
+        internal static string GetStrongHashSuitableForObjectName(string name)
+        {
+            using (MemoryStream ms = new MemoryStream())
+            {
+                using (BinaryWriter b = new BinaryWriter(ms))
+                {
+                    b.Write(name.ToUpperInvariant());
+                }
+
+                ms.Position = 0;
+                return GetStrongHashSuitableForObjectName(ms);
+            }
+        }
+
+        internal static string GetStrongHashSuitableForObjectName(Stream stream)
+        {
+            using (SHA1 sha1 = SHA1.Create())
+            {
+                return ToBase32StringSuitableForDirName(sha1.ComputeHash(stream));
+            }
+        }
+
+        // This is from the NetFX Path class. The implementation in CoreFx was optimized for internal Path usage so
+        // we can't share the implementation.
+        internal static string ToBase32StringSuitableForDirName(byte[] buff)
+        {
+            // This routine is optimised to be used with buffs of length 20
+            Debug.Assert(((buff.Length % 5) == 0), "Unexpected hash length");
+
+            StringBuilder sb = new StringBuilder();
```

Comment:
>StringBuilder sb = new StringBuilder(); [](start = 12, length = 39)

If this will be called often, consider the pattern

StringBuilder sb = StringBuilderCache.Acquire();

return StringBuilderCache.GetStringAndRelease(sb);